src/github_runner_manager/openstack_cloud/health_checks.py

```diff
@@ -0,0 +1,241 @@
+#  Copyright 2024 Canonical Ltd.
+#  See LICENSE file for licensing details.
+
+"""Collection of functions related to health checks for a runner VM."""
+
+import logging
+from datetime import datetime, timedelta
+
+import invoke
+from fabric import Connection as SSHConnection
+
+from github_runner_manager.errors import KeyfileError, SSHError
+from github_runner_manager.manager.cloud_runner_manager import CloudInitStatus, CloudRunnerState
+from github_runner_manager.openstack_cloud.constants import (
+    METRICS_EXCHANGE_PATH,
+    RUNNER_LISTENER_PROCESS,
+    RUNNER_WORKER_PROCESS,
+)
+from github_runner_manager.openstack_cloud.openstack_cloud import OpenstackCloud, OpenstackInstance
+from github_runner_manager.utilities import retry
+
+logger = logging.getLogger(__name__)
+
+_HealthCheckResult = bool | None  # None indicates that the check can not determine health status
+
+
+def check_runner(openstack_cloud: OpenstackCloud, instance: OpenstackInstance) -> bool:
+    """Run health check on a runner.
+
+    Args:
+        openstack_cloud: The OpenstackCloud instance to use
+        instance: The instance hosting the runner to run health check on.
+
+    Returns:
+        True if runner is healthy.
+    """
+    if (check_ok := _health_check_cloud_state(instance)) is not None:
+        return check_ok
+
+    try:
+        ssh_conn = _get_ssh_connection(openstack_cloud=openstack_cloud, instance=instance)
+    except KeyfileError:
+        logger.exception(
+            "Health check failed due to unable to find keyfile for %s", instance.server_name
+        )
+        return False
+    except SSHError:
+        logger.exception("SSH Failed on %s, marking as unhealthy.")
+        return False
+
+    return check_active_runner(ssh_conn, instance)
+
+
+def check_active_runner(
```

Comment:
Do the names `check_runner` and `check_active_runner` make sense @yanksyoon ? The previous names were `runner_health_check` and `run_health_check`, which were not very clear to me.